cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -0,0 +1,402 @@
+      doWithRetry(application, this::retryableDeployApp);
+    }
+  }
+
+  private void retryableDeployApp(SystemApplication application) throws Exception {
+    ApplicationId applicationId = getApplicationId(application);
+    LOG.debug("Deploying app {}", applicationId);
+    try {
+      if (isAppDeployed(applicationId)) {
+        //Already deployed.
+        LOG.debug("Application {} is already deployed", applicationId);
+        return;
+      }
+      String configString = application.getConfig() == null ? null : GSON.toJson(application.getConfig());
+      applicationLifecycleService
+        .deployApp(applicationId.getParent(), applicationId.getApplication(), applicationId.getVersion(),
+                   application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
+    } catch (Exception ex) {
+      checkForRetry(ex);
+      throw new RetryableException(ex);
+    }
+  }
+
+  private boolean isAppDeployed(ApplicationId applicationId) throws Exception {
+    try {
+      applicationLifecycleService.getAppDetail(applicationId);
+      return true;
+    } catch (ApplicationNotFoundException exception) {
+      return false;
+    }
+  }
+
+  //Find all applications for capability and call consumer for each
+  private void doForAllAppsWithMetadata(String capability, CheckedConsumer<ApplicationId> consumer) throws Exception {
+    for (NamespaceMeta namespaceMeta : namespaceAdmin.list()) {
+      int offset = 0;
+      int limit = 100;
+      NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+      EntityResult<ApplicationId> results = getApplications(namespaceId, capability, null,
+                                                            offset, limit);
+      while (!results.getEntities().isEmpty()) {
+        //call consumer for each entity
+        for (ApplicationId entity : results.getEntities()) {
+          consumer.accept(entity);
+        }
+        offset += limit;
+        results = getApplications(namespaceId, capability, results.getCursor(), offset, limit);
+      }
+    }
+  }
+
+  private void stopAllRunningProgramsForApp(ApplicationId applicationId) {
+    try {
+      doWithRetry(applicationId, this::retryableStopRunningPrograms);
+    } catch (Exception ex) {
+      LOG.error("Stopping programs for application {} failed with {}", applicationId, ex);
+    }
+  }
+
+  private void retryableStopRunningPrograms(ApplicationId applicationId) throws Exception {
+    try {
+      programLifecycleService.stopAll(applicationId);
+    } catch (Exception ex) {
+      checkForRetry(ex);
+      throw new RetryableException(ex);
+    }
+  }
+
+  private void deleteAppWithRetry(ApplicationId applicationId) throws Exception {
+    doWithRetry(applicationId, this::retryableDeleteApp);
+  }
+
+  private void retryableDeleteApp(ApplicationId applicationId) throws Exception {
+    try {
+      applicationLifecycleService.removeApplication(applicationId);
+    } catch (Exception ex) {
+      checkForRetry(ex);
```

Comment:
You don't need try catch to determine if it can be retried or not. The `Retrie.callWithRetries` method can take a `Predicate<Exception>` to determine if the exception is retryable or not.

Reply:
Thanks, could remove lot of unnecessary methods.